### DISCUSSION TITLE
MappedAsDataclass lazy relationship auto load all relationship?

### DISCUSSION THREAD

**Discussion by nleononline** [sqlalchemy/sqlalchemy#12179] [Usage Questions]
How to properly work with relationship when using MappedAsDataclass or why lazy relations, automatic loading of all relations?
python: 3.11
sqlalchemy: 2.0.36

**base config:**
```python
async_engine = create_async_engine(
    url=CONFIG.get_asyncpg_url,
    pool_size=10,
    max_overflow=100,
    connect_args={
        "server_settings": {
            "application_name": "Test"
        }
    },
)

async_session_factory = async_sessionmaker(
    bind=async_engine,
    expire_on_commit=False
)



def patch_dataclass(cls, **kwargs):
    kwargs.setdefault("init", False)
    return dataclass(cls, **kwargs)  # type: ignore
    
    
class CreatedUpdatedMixin(MappedAsDataclass, kw_only=True, init=False):
    created_at: Mapped[datetime.datetime] = mapped_column(
        default=None,
        server_default=func.current_timestamp(),
        insert_default=func.now()
    )
    updated_at: Mapped[datetime.datetime] = mapped_column(
        default_factory=datetime.datetime.now,
        init=True,
        server_default=func.current_timestamp(),
        insert_default=func.now(),
        onupdate=func.now(),
        server_onupdate=func.current_timestamp()
    )

class Base(DeclarativeBase, CreatedUpdatedMixin, kw_only=True):
    type_annotation_map = {
        datetime.datetime: TIMESTAMP(timezone=False),
    }
```

**models:**
```python
class UserMapped(Base, kw_only=True, dataclass_callable=patch_dataclass):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True, default=None)
    name: Mapped[str]
    status: Mapped[str] = mapped_column(default="ACTIVE", server_default="ACTIVE")
    is_paid: Mapped[bool] = mapped_column(default=False, server_default="false")

    settings: Mapped[UserSettingsMapped] = relationship(
        argument="UserSettingsMapped", back_populates="user",
        lazy="selectin",  << selectin
        uselist=False, default=None
    )


class UserSettingsMapped(Base, kw_only=True, dataclass_callable=patch_dataclass):
    __tablename__ = "user_settings"

    user_id: Mapped[int] = mapped_column(
        ForeignKey("user.id", ondelete="CASCADE"),
        primary_key=True, default=None
    )

    is_admin: Mapped[bool] = mapped_column(default=False, server_default="false")
    first_paid_dt: Mapped[datetime.datetime | None] = mapped_column(default=None)

    user: Mapped[UserMapped] = relationship(
        argument="UserMapped", back_populates="settings",
        lazy="selectin",  << selectin
        uselist=False, default=None
    )
```

**main.py:**
```python
async def main():
    async with async_session_factory() as session:
        session: AsyncSession
        user = UserMapped(
            id=2,
            name="Test User"
        )
        user = await session.merge(user)
        await session.flush([user])
        await session.commit()
        print(user.__dict__)
```

UserSettingsMapped is in user, although there was no access to the attribute. Moreover, UserSettingsMapped also has its own UserMapped...
```python
>> {
    '_sa_instance_state': <sqlalchemy.orm.state.InstanceState object at 0x103815c70>, 
    'is_paid': False, 
    'status': 'ACTIVE', 
    'updated_at': datetime.datetime(2024, 12, 10, 13, 6, 11, 756539), 
    'id': 2, 
    'name': 'Test User', 
    'created_at': datetime.datetime(2024, 12, 10, 13, 6, 11, 756539),
    'settings': UserSettingsMapped(
        created_at=datetime.datetime(2024, 12, 10, 13, 6, 11, 756539), 
        updated_at=datetime.datetime(2024, 12, 14, 15, 17, 23, 918717), 
        user_id=2, 
        is_admin=False, 
        first_paid_dt=datetime.datetime(2024, 12, 14, 15, 17, 23, 889997), 
        user=UserMapped(
            created_at=datetime.datetime(2024, 12, 10, 13, 6, 11, 756539), 
            updated_at=datetime.datetime(2024, 12, 10, 13, 6, 11, 756539), 
            id=2, 
            name='Test User', 
            status='ACTIVE', 
            is_paid=False, 
            settings=...
        )
    )
}
```

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you are asking here.
The relationships are loaded because you set them to eagerly load using `selectin`. Don't do that if you don't want to eagerly load them

**Reply by nleononline** [reply]
> Привет,
> 
> Я не уверен, о чем вы здесь спрашиваете. Связи загружены, потому что вы установили их для активной загрузки с помощью `selectin`. Не делайте этого, если вы не хотите, чтобы они были загружены активно

If you specify `lazy="noload"`, an error occurs when used with session.merge()

```python
    async with async_session_factory() as session:
        session: AsyncSession
        user = UserMapped(
            id=2,
            name="Test User",
            settings=UserSettingsMapped(
                user_id=2,
                is_admin=True,
                first_paid_dt=datetime.datetime.now(),
            )
        )
        user = await session.merge(user)
        await session.flush([user])
        await session.commit()
        print(user)
        
        
error:
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)
```

An error occurs when calling `print(user)`

**Reply by CaselIT** [reply]
Since you are using dataclasses, the repr will access every field by default, so they need to bo loaded otherwise you get an error when using asyncio since you can do IO implicitly. Usual solution is to use repr=False for relationships.

Regarding merge, not sure. Have you tried leaving the lazy to its default value?

**Reply by nleononline** [reply]
> Поскольку вы используете dataclasses, repr будет иметь доступ к каждому полю по умолчанию, поэтому они должны быть загружены, в противном случае вы получите ошибку при использовании asyncio, поскольку вы можете сделать IO неявно. Обычное решение — использовать repr=False для отношений.
> 
> Насчет merge не уверен. Пробовали ли вы оставить lazy в значении по умолчанию?

Specifying `repr=False` does solve the problem, but not completely.

Technically, after calling `session.merge()` after merging, the current data is returned from the database and the current data is in the object (I can get it), but why an exception is thrown when calling `print(model)`... why?

**default `repr`:**
```python
async def main():
    async with async_session_factory() as session:
        session: AsyncSession
        user = UserMapped(
            id=2,
            name="Test User",
            settings=UserSettingsMapped(
                user_id=2,
                # is_admin=False
            )
        )
        user = await session.merge(user)
        await session.flush([user])
        await session.commit()

        print(user.settings.is_admin)  << False (None in object before the call session.merge())
        print(user.settings) << error sqlalchemy.exc.MissingGreenlet
```

If you specify `repr=False` and `lazy="noload"`, the `sqlalchemy.exc.MissingGreenlet` error disappears
```python
        print(user.settings.is_admin)  << False
        print(user.settings) << relation object
```

but if you change any value of the associated object and try to access the associated object, the error is returned, because `lazy="nolad"` does not load related objects, despite the fact that you passed the related object; 
```python
async def main():
    async with async_session_factory() as session:
        session: AsyncSession
        user = UserMapped(
            id=2,
            name="Test User",
            settings=UserSettingsMapped(
                user_id=2,
                is_admin=True   << database: is_admin=False
            )
        )
        user = await session.merge(user)
        await session.flush([user])
        await session.commit()

        print(user.settings.is_admin)  << error sqlalchemy.exc.MissingGreenlet
        print(user.settings) << error sqlalchemy.exc.MissingGreenlet
```

if the data has not been updated, it will load them again (I use `init=False`, some of the data is empty and they are filled with data from the database)

```python
async def main():
    async with async_session_factory() as session:
        session: AsyncSession
        user = UserMapped(
            id=2,
            # name="Test User"
        )
        user = await session.merge(user)
        await session.flush([user])
        await session.commit()

        print(user.name)  << "Test User" (None in object before the call session.merge())
```

---

The default is `lazy="select"`, but this does not work for the asyncio operation; the same exception is thrown as above. In this regard, I don’t understand why `lazy="noload"` is not specified as standard. 

Since we're talking about this, I'll add that the need to specify `uselist=False` when using mapping with dataclass looks extremely strange, because `uselist` is specified as standard `uselist: bool | None = None`, and `None` is expected to be `False`, but in mapping with dataclass it behaves as `True`. I somehow spent a whole day trying to understand this special thing, and it really would be worth fixing.

**Reply by CaselIT** [reply]
> The default is lazy="select", but this does not work for the asyncio operation; the same exception is thrown as above. In this regard, I don’t understand why lazy="noload" is not specified as standard.

noload should be used only in special cases, since it's returning data that's not correct. Select has been the default since before asyncio.

> but why an exception is thrown when calling print(model)... why?

It may have unloaded relationship in some object.

Regarding the test, it's really hard to follow since you haven't provided a runnable script, so I don't know how the session returned from `async_session_factory` is configued. You are committing before doing your test, so you have to take into consideration this flag https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.expire_on_commit

> Since we're talking about this, I'll add that the need to specify uselist=False when using mapping with dataclass looks extremely strange, because uselist is specified as standard uselist: bool | None = None, and None is expected to be False, but in mapping with dataclass it behaves as True. I somehow spent a whole day trying to understand this special thing, and it really would be worth fixing.

Can you provide a runnable script that shows this? It may well be a bug.

**Reply by nleononline** [reply]
> noload следует использовать только в особых случаях, так как он возвращает неверные данные.

what does invalid data mean? 0_о

The question also arises: what to use with MappedAsDataclass in this case? 
After all, any loading strategy loads ALL relationships, even if you did not access them, because... When converting an object to a dataclass, the attribute is actually accessed.

> Возможно, в каком-то объекте есть незагруженные связи.

There are no other relationships in the object. This is a test object, it is described in the first comment.


> Что касается теста, то его действительно трудно контролировать, так как вы не включаете работающий скрипт, поэтому я не знаю, как происходит сеанс, возвращаемый из `async_session_factory`

I thought it was obvious and left it out. I've updated my first comment to include all the missing session settings.

**Reply by CaselIT** [reply]
> what does invalid data mean? 0_о

noload meas noload, meaning that you get no data regardless of what the db has. So in a relationship you get none / empty list regardless. See https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#sqlalchemy.orm.noload

I've tried a bit your examples and they seem to work locally, but maybe I don't have the right data in the db. Can you post one that shows the error starting from an empty db?

**Reply by nleononline** [reply]
database: psql

`src/database.py`:
```python
import datetime
from dataclasses import dataclass

from sqlalchemy import TIMESTAMP, func
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, MappedAsDataclass

async_engine = create_async_engine(
    url="YOUR URL",  # TODO: set your value
    pool_size=10,
    max_overflow=100,
    connect_args={
        "server_settings": {
            "application_name": "Test"
        }
    },
)

async_session_factory = async_sessionmaker(
    bind=async_engine,
    expire_on_commit=False
)


def patch_dataclass(cls, **kwargs):
    kwargs.setdefault("init", False)
    return dataclass(cls, **kwargs)  # type: ignore


class CreatedUpdatedMixin(MappedAsDataclass, kw_only=True, init=False):
    created_at: Mapped[datetime.datetime] = mapped_column(
        default=None,
        server_default=func.current_timestamp()
    )
    updated_at: Mapped[datetime.datetime] = mapped_column(
        default=None,
        server_default=func.current_timestamp(),
        onupdate=func.now(),
        server_onupdate=func.current_timestamp()
    )


class Base(DeclarativeBase, CreatedUpdatedMixin, kw_only=True, init=False):
    type_annotation_map = {
        datetime.datetime: TIMESTAMP(timezone=False),
    }

```

`models.__init__.py`
```python
from .user import UserMapped
from .user_settings import UserSettingsMapped

__all__ = ["UserMapped", "UserSettingsMapped"]
```

`models/user.py`
```python
from __future__ import annotations
from typing import TYPE_CHECKING

from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.database import Base, patch_dataclass

if TYPE_CHECKING:
    from src.models import UserSettingsMapped


class UserMapped(Base, kw_only=True, dataclass_callable=patch_dataclass):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True, default=None)
    name: Mapped[str]
    status: Mapped[str] = mapped_column(default="ACTIVE", server_default="ACTIVE")
    is_paid: Mapped[bool] = mapped_column(default=False, server_default="false")

    settings: Mapped[UserSettingsMapped] = relationship(
        argument="UserSettingsMapped", back_populates="user",
        lazy="noload",
        uselist=False, default=None, repr=False
    )

```


`models/user_settings.py`:
```python
from __future__ import annotations
from typing import TYPE_CHECKING
import datetime

from sqlalchemy import ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.database import Base, patch_dataclass

if TYPE_CHECKING:
    from src.models import UserMapped


class UserSettingsMapped(Base, kw_only=True, dataclass_callable=patch_dataclass):
    __tablename__ = "user_settings"

    user_id: Mapped[int] = mapped_column(
        ForeignKey("user.id", ondelete="CASCADE"),
        primary_key=True, default=None
    )

    is_admin: Mapped[bool] = mapped_column(default=False, server_default="false")
    first_paid_dt: Mapped[datetime.datetime | None] = mapped_column(default=None)

    user: Mapped[UserMapped] = relationship(
        argument="UserMapped", back_populates="settings",
        lazy="noload",
        uselist=False, default=None, repr=False
    )

```

`src/main.py`:
```python
import asyncio
import datetime

from src.database import async_session_factory
from src.models import UserMapped, UserSettingsMapped


async def create_objects():
    async with async_session_factory() as session:
        user = UserMapped(
            id=1,
            name="Test User",
            is_paid=True,
            settings=UserSettingsMapped(
                user_id=1,
                is_admin=True,
                first_paid_dt=datetime.datetime.now(),
            )
        )
        user = await session.merge(user)
        await session.flush([user])
        await session.commit()

        assert user.id == 1
        assert user.name == "Test User"
        assert user.status == "ACTIVE"
        assert user.is_paid is True
        assert isinstance(user.created_at, datetime.datetime)
        assert isinstance(user.updated_at, datetime.datetime)

        assert user.settings.user_id == 1
        assert user.settings.is_admin is True
        assert isinstance(user.settings.first_paid_dt, datetime.datetime)
        assert isinstance(user.settings.created_at, datetime.datetime)
        assert isinstance(user.settings.updated_at, datetime.datetime)
        print("create_objects success")


async def update_objects():
    async with async_session_factory() as session:
        user = UserMapped(
            id=1,
            name="Test User",
            status="STOP",
            settings=UserSettingsMapped(
                user_id=1,
                is_admin=False
            )
        )
        user = await session.merge(user)
        await session.flush([user])
        await session.commit()

        assert user.id == 1
        assert user.name == "Test User"
        assert user.status == "STOP"
        assert user.is_paid is True
        assert isinstance(user.created_at, datetime.datetime)
        assert isinstance(user.updated_at, datetime.datetime)

        assert user.settings.user_id == 1
        assert user.settings.is_admin is False
        assert isinstance(user.settings.first_paid_dt, datetime.datetime)
        assert isinstance(user.settings.created_at, datetime.datetime)
        assert isinstance(user.settings.updated_at, datetime.datetime) # ERROR sqlalchemy.exc.MissingGreenlet WTF??? 
        print("update_objects success")


async def main():
    await create_objects()
    await update_objects()


if __name__ == "__main__":
    asyncio.run(main())
```

**Reply by CaselIT** [reply]
thanks for creating a full example.

the updated_at column has a server_onupdate / onupdate clause. By default sqlalchemy does not fetch back those changed values, unless eager defaults is enabled. see https://docs.sqlalchemy.org/en/20/orm/persistence_techniques.html#orm-server-defaults

if you do something like this this example works. 
```py
class Base(DeclarativeBase, CreatedUpdatedMixin, kw_only=True, init=False):
    type_annotation_map = {
        datetime.datetime: TIMESTAMP(timezone=False),
    }
    __mapper_args__ = {"eager_defaults": True}
```

@zzzeek I don't remember if in 2.1 we had in plan of chaning this so that it's automatic if the driver has returning support or not

**Reply by CaselIT** [reply]
> @zzzeek I don't remember if in 2.1 we had in plan of chaning this so that it's automatic if the driver has returning support or not

we just discussed it I think at some point but there was no issue. I've just created one: https://github.com/sqlalchemy/sqlalchemy/issues/12182
